core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java

```diff
@@ -372,24 +379,31 @@ public ClusterState execute(ClusterState currentState) {
                 }
 
                 nodesBuilder = DiscoveryNodes.builder(currentState.nodes());
-                Iterator<Map.Entry<DiscoveryNode, List<MembershipAction.JoinCallback>>> iterator = pendingJoinRequests.entrySet().iterator();
+                Iterator<Map.Entry<DiscoveryNode, List<MembershipAction.JoinCallback>>> iterator =
+                    pendingJoinRequests.entrySet().iterator();
                 while (iterator.hasNext()) {
                     Map.Entry<DiscoveryNode, List<MembershipAction.JoinCallback>> entry = iterator.next();
                     final DiscoveryNode node = entry.getKey();
-                    joinCallbacksToRespondTo.addAll(entry.getValue());
                     iterator.remove();
-                    if (currentState.nodes().nodeExists(node.getId())) {
+                    final String preflight = nodesBuilder.preflightPut(node);
+                    if (preflight != null) {
+                        Throwable failure = new IllegalStateException(preflight);
+                        for (MembershipAction.JoinCallback callback: entry.getValue()) {
+                            joinCallbacksToFail.add(new Tuple<>(callback, failure));
+                        }
+                    } else if (currentState.nodes().nodeExists(node.getId())) {
                         logger.debug("received a join request for an existing node [{}]", node);
+                        joinCallbacksToRespondTo.addAll(entry.getValue());
```

Comment:
hmm.. you're right. A quick restart + change of attributes can make this happen. I'll add this and add a test.